HashStore-java storeObject refactor

doulikecookiedough opened the pull request:
Hi @taojing2002 and @artntek,

When you guys have scope, can you please help code review HashStore-java thus far? I would like to ensure that my code is clean and following good practices before I proceed any further. Thank you!

artntek commented on src/test/java/org/dataone/hashstore/HashStoreTest.java:
General - all tests: For JUnit4 with the `@Test` annotation, we no longer need to start a test name with `test___`. It increases readability and search compatibility if you simply use the same name as the method you're testing.

When you have multiple test cases for the same method, it might be clearer to do what [the Google code style doc suggests](https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names), and start the method name with the name of the method under test, followed by an underscore, and then the conditions for this particular test:
_

> "Underscores may appear in JUnit test method names to separate logical components of the name, with each component written in [lowerCamelCase], for example transferMoney_deductsFromSource. There is no One Correct Way to name test methods."
_
So for example, your tests might be called:
```
storeObject_relPath()
storeObject_absPath()
storeObject_moveIsDuplicate()
etc
```
which seems more readable.

doulikecookiedough replied:
Updating all test names! Thank you for clarifying.

artntek commented on src/test/java/org/dataone/hashstore/HashStoreTest.java:
```diff
@@ -1,18 +1,351 @@
+ * Unit tests for HashStore-java
  */
 public class HashStoreTest {
+    public HashStore hsj;
```
Less confusing to call this `hashStore`, instead of `hsj`

artntek commented on src/test/java/org/dataone/hashstore/HashStoreTest.java:
```diff
@@ -1,18 +1,351 @@
+                HashAddress objInfo = hsj.storeObject(dataStream, pid, null, null, null);
+            } catch (Exception e) {
+                Class<?> exceptionClass = e.getClass();
+                assertTrue(exceptionClass == FileAlreadyExistsException.class);
```
can replace 2 lines with
`assertTrue(e instanceof FileAlreadyExistsException);`

doulikecookiedough replied:
Thank you - this is way cleaner!

artntek commented on src/test/java/org/dataone/hashstore/HashStoreTest.java:
```diff
@@ -1,18 +1,351 @@
+                HashAddress objInfoDup = hsj.storeObject(dataStreamDup, pid, null, null, null);
+            } catch (Exception e) {
+                Class<?> exceptionClass = e.getClass();
+                assertTrue(exceptionClass == FileAlreadyExistsException.class);
```
assertTrue(e instanceof FileAlreadyExistsException);

artntek commented on src/test/java/org/dataone/hashstore/hashfs/HashUtilTest.java:
```diff
@@ -25,41 +22,89 @@
+    public File generateTemporaryFile() throws Exception {
         String prefix = "testfile";
         File directory = tempFolder.getRoot();
         File newFile = null;
```
redundant - could just assign in next line

artntek commented on src/test/java/org/dataone/hashstore/hashfs/HashFileStoreTest.java:
```diff
@@ -45,174 +55,336 @@ public void initializeHashFileStore() {
+    /**
+     * Verify exception thrown when checksum provided does not match
+     */
+    public void testPutCorrectChecksumValue() throws Exception {
```
missing @Test annotation

artntek commented on src/test/java/org/dataone/hashstore/hashfs/HashFileStoreTest.java:
```diff
@@ -23,18 +25,26 @@
+    public String tmpStringFull;
 
+    public TestDataHarness testData = new TestDataHarness();
+    public HashUtil hsil = new HashUtil();
```
`hashUtil` is less confusing than `hsil`

artntek commented on src/test/java/org/dataone/hashstore/hashfs/HashFileStoreTest.java:
```diff
@@ -45,174 +55,336 @@ public void initializeHashFileStore() {
-    public void testDefaultStoreDirectory() {
-        try {
-            HashFileStore defaultHfs = new HashFileStore(3, 2, "SHA-256", null);
+    public void testDefaultStoreDirectoryNull() throws Exception {
```
see review note on HashStoreTest about naming. this one's tough to decipher :)

artntek reviewed (changes requested):
looking good - some really nice work here, and great test coverage! I started marking the unused variables, but stopped after we talked on Slack - I understand you'll be mopping those up with your VSCode inspector.

doulikecookiedough commented on src/main/java/org/dataone/hashstore/HashStore.java:
Refactored HashStore and HashFileStore classes and junit tests

doulikecookiedough commented on src/main/java/org/dataone/hashstore/HashStore.java:
```diff
@@ -30,6 +39,95 @@ public HashStore(String storeDirectory)
+    /**
+     * Store an object to the storeDirectory.
+     * 
+     * The permanent address is the SHA256 hex digest of a given string that
```
Fixed typo

doulikecookiedough commented on src/main/java/org/dataone/hashstore/HashStore.java:
```diff
@@ -30,6 +39,95 @@ public HashStore(String storeDirectory)
+     * Store an object to the storeDirectory.
+     * 
+     * The permanent address is the SHA256 hex digest of a given string that
+     * represents an authority based identifer (ex. pid)
```
Fixed typo

doulikecookiedough commented on src/main/java/org/dataone/hashstore/HashStore.java:
```diff
@@ -30,6 +39,95 @@ public HashStore(String storeDirectory)
+        if (object == null) {
+            throw new NullPointerException("Invalid input stream, data is null.");
+        }
+        if (pid == null || pid == "") {
```
Thank you! Updated

doulikecookiedough commented on src/main/java/org/dataone/hashstore/HashStore.java:
```diff
@@ -14,6 +20,9 @@ public class HashStore {
     private String sysmetaNameSpace = "http://ns.dataone.org/service/types/v2.0";
```
It will be used soon in (storeSysmeta), but noted that I will not add unused variables until I need them in the future.

doulikecookiedough commented on src/main/java/org/dataone/hashstore/HashStore.java:
```diff
@@ -30,6 +39,95 @@ public HashStore(String storeDirectory)
+        // Lock pid for thread safety, transaction control and atomic writing
+        // A pid can only be stored once and only once, subsequent calls will
+        // be accepted but will be rejected if pid hash object exists
+        synchronized (objectLockedIds) {
```
Noted, thank you!

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashFileStore.java:
```diff
@@ -36,7 +44,7 @@ public HashFileStore(int depth, int width, String algorithm, String storeDirecto
             throw new IllegalArgumentException("Depth and width must be greater than 0.");
         }
-        if (algorithm == null || algorithm.isEmpty()) {
+        if (algorithm == null || algorithm == "") {
```
I felt like I had to be extra specific at that time - but I agree with your recommendation and have reverted it where applicable. Thank you!

taojing2002 commented on src/main/java/org/dataone/hashstore/HashStore.java:
To me, HashStore should be an abstract class or interface. The  HashFileStore class should extend or implement it.

doulikecookiedough replied:
I agree with you here - and I have created a new issue to refactor the code as such: [Refactor Hashstore to be an abstract class or interface](https://github.com/DataONEorg/hashstore-java/issues/14)

taojing2002 commented on src/main/java/org/dataone/hashstore/hashfs/HashUtil.java:
Will those methods be shared by more than one class? All of them can be in the HashFileStore class.

doulikecookiedough replied:
Will address during abstract class/interface refactor: [Refactor Hashstore to be an abstract class or interface](https://github.com/DataONEorg/hashstore-java/issues/14)

taojing2002 commented on src/main/java/org/dataone/hashstore/HashStore.java:
To me, the `put` method seems duplicated. You may put its implementation into the `putObject` method.

doulikecookiedough replied:
Will address during abstract class/interface refactor: [Refactor Hashstore to be an abstract class or interface](https://github.com/DataONEorg/hashstore-java/issues/14)

taojing2002 commented on src/main/java/org/dataone/hashstore/hashfs/HashFileStore.java:
In line 151, you need to consider the situations when `checksumAlgorithm` is empty or  `checksum` is null.

taojing2002 commented on src/main/java/org/dataone/hashstore/hashfs/HashFileStore.java:
```diff
@@ -46,91 +54,162 @@ public HashFileStore(int depth, int width, String algorithm, String storeDirecto
 
-        // Validate object if algorithm and checksum is passed
+        // Validate object if checksum and checksum algorithm is passed
         if (additionalAlgorithm != null && checksum != null) {
```
This line confuses me.  Is the `checksum` is for  `additionalAlgorithm` or `checksumAlgorithm`?

doulikecookiedough replied:
It was a mistake, I have addressed it in my latest commits. It should be for checksumAlgorithm.

taojing2002 commented:
The `HashAddress` class doesn't have any documentation. Is it possible to rename the class name?

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashFileStore.java:
I will update these and all omissions once I've had a chance to discuss the structure/set-up with @taojing2002 . He mentioned "HashStore should be an abstract class or interface. The HashFileStore class should extend or implement it."

Given our previous discussions about how we may want to swap the backend out for HashStore in the future, I want to update these classes as per Jing's suggestion.

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashUtil.java:
```diff
@@ -88,26 +95,67 @@ public String shard(int depth, int width, String digest) {
+        }
+        if (!algorithmSupported) {
+            throw new IllegalArgumentException(
+                    "Algorithm not supported. Supported algorithms: " + this.supportedHashAlgorithms);
```
Updated - thank you for the clarification!

taojing2002 commented on src/main/java/org/dataone/hashstore/hashfs/HashUtil.java:
```diff
@@ -44,27 +47,31 @@ public boolean validateAlgorithm(String algorithm) {
      */
-    public File generateTmpFile(String prefix, File directory) throws IOException {
+    public File generateTmpFile(String prefix, File directory) throws IOException, SecurityException {
         String newPrefix = prefix + "-" + System.currentTimeMillis();
```
If two threads create temp files at the same time, they may get the same file name. So this causes collision.

doulikecookiedough replied:
Thank you - I added an additional random number:
```
Random random = new Random();
int randomNumber = random.nextInt(900000);
String newPrefix = prefix + "-" + System.currentTimeMillis() + randomNumber;
```

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashUtil.java:
```diff
@@ -88,26 +95,67 @@ public String shard(int depth, int width, String digest) {
-        if (!algorithmSupported) {
+            String additionalAlgorithm)
+            throws NoSuchAlgorithmException, IOException, FileNotFoundException, SecurityException {
+        if (additionalAlgorithm == "") {
```
Got it - thank you!

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashUtil.java:
Removed the if statement

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashUtil.java:
```diff
@@ -44,27 +47,31 @@ public boolean validateAlgorithm(String algorithm) {
-    public File generateTmpFile(String prefix, File directory) throws IOException {
+    public File generateTmpFile(String prefix, File directory) throws IOException, SecurityException {
         String newPrefix = prefix + "-" + System.currentTimeMillis();
         String suffix = null;
```
Yes - this is intentional. I would like for the createTempFile method to proceed with what the doc outlines - "The suffix argument may be null, in which case the suffix ".tmp" will be used."

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashFileStore.java:
```diff
@@ -1,8 +1,10 @@
 package org.dataone.hashstore.hashfs;
 
 import java.io.File;
```
I will use java.nio.file.Files (and its related class and class methods) where possible - thank you!

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashFileStore.java:
```diff
@@ -46,91 +54,162 @@ public HashFileStore(int depth, int width, String algorithm, String storeDirecto
-
         // If no path provided, create default path with user.dir root + /HashFileStore
-        if (storeDirectory == null) {
+        if (storeDirectory == null || storeDirectory == "") {
```
Deleted or condition after changing my variable for paths from String to Path objects

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashUtil.java:
```diff
@@ -44,27 +47,31 @@ public boolean validateAlgorithm(String algorithm) {
     /**
      * Create a list of 'depth' number of tokens with 'width' with the last item
-     * being the remainder of the digest, and return a String path
+     * being the remainder of the digest, delimited by "/"
```
Updated the javadoc to "Generates a hierarchical path by dividing a given digest into tokens of fixed width, and concatenating them with '/' as the delimiter."

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashUtil.java:
```diff
@@ -88,26 +95,67 @@ public String shard(int depth, int width, String digest) {
         }
-        String shardedPath = "/" + String.join("/", stringArray);
-        return shardedPath;
+        String stringShard = "/" + String.join("/", stringArray);
```
It seems like compatibility issues continue to be a challenge. I added "/" in the beginning here so that the code where I call it would look cleaner - but this makes the method messy, so I removed "/" here and appended it where applicable in HashFileStore (ex. `this.objectStoreDirectory.toString() + "/" + objShardString;`)

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashUtil.java:
```diff
@@ -88,26 +95,67 @@ public String shard(int depth, int width, String digest) {
+     */
+    public String getHexDigest(String string, String algorithm) throws NoSuchAlgorithmException {
+        boolean algorithmSupported = this.validateAlgorithm(algorithm);
+        if (algorithm == null || algorithm == "" | string == null || string == "") {
```
This was a bug/error on my part. I refactored to split the check for algorithm and string for clarity.

doulikecookiedough commented on src/main/java/org/dataone/hashstore/hashfs/HashUtil.java:
```diff
@@ -167,14 +215,15 @@ protected Map<String, String> writeToTmpFileAndGenerateChecksums(File tmpFile, I
      * @throws IOException
      */
-    protected boolean move(File source, File target) throws IOException {
+    protected boolean move(File source, File target) throws IOException, SecurityException {
```
I've updated move to reflect your comments - refactored HashFileStore/HashAddress and its tests where applicable (after renaming `isDuplicate` variable to `isNotDuplicate`)

doulikecookiedough commented:
Thank you @artntek and @taojing2002 for your feedback. I believe I have addressed all points and moved any outstanding discussion/action items to: [Refactor Hashstore to be an abstract class or interface](https://github.com/DataONEorg/hashstore-java/issues/14). I am merging this pull request and will continue in a new branch.
